File: src/main/java/net/adoptium/model/Download.java
package net.adoptium.model;

import net.adoptopenjdk.api.v3.models.Binary;

import java.util.Objects;

/**
 * A Download is identified by a binary and a version (semver)
 * This is what we effectively pass to /thank-you/{binary}
 */
public class Download {
    private final Binary binary;
    private final String semver;

    public Download(Binary binary, String semver) {
        this.binary = binary;
        this.semver = semver;
    }

    public Binary getBinary() {
        return binary;
    }

    public String getSemver() {
        return semver;
    }

    @Override
    public String toString() {
        return "Download{" +
                "binary=" + binary +
                ", semver='" + semver + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Download download = (Download) o;
        return binary.equals(download.binary) && semver.equals(download.semver);
    }

    @Override
    public int hashCode() {
        return Objects.hash(binary, semver);
    }
}

File: src/main/kotlin/net/adoptium/model/Download.kt
package net.adoptium.model

// Download represents an abstract entity which a client can downlaod
// It's usually a runtime or jdk
// TODO use adoptopenjdk-api-v3-models, name = JvmImpl
data class Download(
        // human-readable
        val name: String,
        // for api
        val codename: String,

        // human-readable
        val version: String,
        // for api
        val versionCode: String,

        val url: String
) {
    companion object {
        // TODO import OpenJDK V3 API, build Download from their Entities
        /*fun fromRuntime(runtime: Runtime): Download {
            return Download(runtime.);
        }*/
    }
}
